### PR TITLE
Add implicit coersions to VALUES

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

**Comment by dain** [comment]
fixes #2198

**Comment by martint** [comment]
Typo in commit message: coersions -> coercions

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java [hunk 1]
```diff
@@ -539,32 +541,60 @@ protected TupleDescriptor visitValues(Values node, AnalysisContext context)
...
 
-        if (types.size() > 1) {
-            throw new SemanticException(MISMATCHED_SET_COLUMN_TYPES,
-                    node,
-                    "Values rows have mismatched types: %s vs %s",
-                    Iterables.get(types, 0),
-                    Iterables.get(types, 1));
+        // determine common super type of the rows
+        List<Type> fieldTypes = new ArrayList<>(rowTypes.iterator().next());
+        for (List<Type> rowType : rowTypes) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java]: You can problaby do this with `rowTypes.stream().reduce(...)`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java [hunk 2]
```diff
@@ -539,32 +541,60 @@ protected TupleDescriptor visitValues(Values node, AnalysisContext context)
...
+                Type expectedType = fieldTypes.get(0);
+                if (!analysis.getType(row).equals(expectedType)) {
+                    analysis.addCoercion(row, expectedType);
+                }
+            }
         }
 
-        TupleDescriptor descriptor = new TupleDescriptor(fields);
+        TupleDescriptor descriptor = new TupleDescriptor(fieldTypes.stream()
+                .map((valueType) -> Field.newUnqualified(Optional.empty(), valueType))
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java]: Parenthesis around `valueType` not needed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 3]
```diff
@@ -179,7 +185,12 @@ protected RelationPlan visitSampledRelation(SampledRelation node, Void context)
         if (node.getType() == SampledRelation.Type.POISSONIZED) {
             sampleWeightSymbol = symbolAllocator.newSymbol("$sampleWeight", BIGINT);
         }
-        PlanNode planNode = new SampleNode(idAllocator.getNextId(), subPlan.getRoot(), ratio, SampleNode.Type.fromType(node.getType()), node.isRescaled(), Optional.ofNullable(sampleWeightSymbol));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 4]
```diff
@@ -190,7 +201,7 @@ protected RelationPlan visitJoin(Join node, Void context)
         RelationPlan leftPlan = process(node.getLeft(), context);
 
         // Convert CROSS JOIN UNNEST to an UnnestNode
-        if (node.getRight() instanceof Unnest || (node.getRight() instanceof  AliasedRelation  && ((AliasedRelation) node.getRight()).getRelation() instanceof Unnest)) {
+        if (node.getRight() instanceof Unnest || (node.getRight() instanceof AliasedRelation && ((AliasedRelation) node.getRight()).getRelation() instanceof Unnest)) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -308,7 +319,9 @@ else if (Iterables.all(firstDependencies, right.canResolvePredicate()) && Iterab
         }
         Optional<Symbol> sampleWeight = Optional.empty();
         if (leftPlanBuilder.getSampleWeight().isPresent() || rightPlanBuilder.getSampleWeight().isPresent()) {
-            Expression expression = new ArithmeticBinaryExpression(ArithmeticBinaryExpression.Type.MULTIPLY, oneIfNull(leftPlanBuilder.getSampleWeight()), oneIfNull(rightPlanBuilder.getSampleWeight()));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 6]
```diff
@@ -494,27 +508,61 @@ else if (type instanceof MapType) {
...
+                    "Constant expression cannot contain as sub-query");
+            analyzer.analyze(canonicalized, new TupleDescriptor(), new AnalysisContext());
 
             // evaluate the expression
-            Object result = ExpressionInterpreter.expressionInterpreter(canonicalized, metadata, session, analysis.getTypes()).evaluate(0);
+            Object result = ExpressionInterpreter.expressionInterpreter(canonicalized, metadata, session, analyzer.getExpressionTypes()).evaluate(0);
             checkState(!(result instanceof Expression), "Expression interpreter returned an unresolved expression");
 
-            return LiteralInterpreter.toExpression(result, analysis.getType(expression));
+            return LiteralInterpreter.toExpression(result, type);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Instead of passing `type` to this method, return `result` and let callers invoke `LiteralInterpreter.toExpression`

**Comment by martint** [comment]
looks good
